### PR TITLE
[clang-tidy] Create bugprone-bitwise-pointer-cast check

### LINKED ISSUE #106987: [clang-tidy] Check request: detect incorrect usage of `std::bit_cast`
`std::bit_cast` is often presented as the correct alternative to type punning in C++. However, we see people use it as a drop-in replacement to `reinterpret_cast` , which only obfuscates the code and still has UB:

```cpp
float x{};
-int* y = reinterpret_cast<int*>(&x);
+int* y = std::bit_cast<int*>(&x);
```

This usage is incorrect. Instead, the code should do:
```cpp
int y = std::bit_cast<int>(x);
```

Most likely all usages of `std::bit_cast<PointerType>` are incorrect.

### PR DESCRIPTION
**Pull request by carlosgalvezp**
To detect unsafe usages of casting a pointer to another via copying
the bytes from one into the other, either via std::bit_cast or via
memcpy.use of bit_cast. This is currently not caught by any other
means.

Fixes #106987

### CODE REVIEW

**Review by HerrCai0907 — COMMENTED** [review]
Should we forbidden bit cast pointer to number and number to pointer? or we should only forbidden bit cast pointer to pointer.

**Comment by carlosgalvezp** [comment]
Perhaps we can start with only pointer-to-pointer, yes. The original intention was to forbid that:

https://www.open-std.org/jtc1/sc22/WG21/docs/papers/2016/p0476r0.html#det

```cpp

**Comment by carlosgalvezp** [comment]
Perhaps we can start with pointer-to-pointer only, yes. Pointer-int conversions are still implementation-defined, but I guess it's less of a problem than UB.

The original paper only checked that both inputs are not pointers:

```cpp
  !(is_pointer_v<From> &&
    is_pointer_v<To>) &&
```

File: clang-tools-extra/docs/clang-tidy/checks/bugprone/bit-cast-pointers.rst [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+The solution to safe type punning is to apply ``std::bit_cast`` on value types,
+not on pointer types:
+
+.. code-block:: c++
+
+    int x{};
+    float y = std::bit_cast<float>(x);
+
+This way, the bytes of the input object are copied into the output object, which
+is safe from Undefined Behavior. Compilers are able to optimize this copy and
```
Comment (5chmidti) [inline clang-tools-extra/docs/clang-tidy/checks/bugprone/bit-cast-pointers.rst:35]: > safe from undefined Behavior

Seems to be not technically the case, according to https://en.cppreference.com/w/cpp/numeric/bit_cast

> If there is no value of type To corresponding to the value representation produced, the behavior is undefined.

> <c++26 For each bit in the value representation of the result that is indeterminate, the smallest object containing that bit has an indeterminate value; the behavior is undefined unless that object is of an [uninitialized-friendly type](https://en.cppreference.com/w/cpp/language/default_initialization#Special_cases).

> \>=c++26 if [bit] b is indeterminate, the behavior is undefined.

I think `[much] safer` would be correctly qualified.

File: clang-tools-extra/clang-tidy/bugprone/BitCastPointersCheck.h [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+class BitCastPointersCheck : public ClangTidyCheck {
+public:
+  BitCastPointersCheck(StringRef Name, ClangTidyContext *Context)
+      : ClangTidyCheck(Name, Context) {}
+  void registerMatchers(ast_matchers::MatchFinder *Finder) override;
+  void check(const ast_matchers::MatchFinder::MatchResult &Result) override;
+  bool isLanguageVersionSupported(const LangOptions &LangOpts) const override {
+    return LangOpts.CPlusPlus20;
+  }
+};
```
Comment (5chmidti) [inline clang-tools-extra/clang-tidy/bugprone/BitCastPointersCheck.h:30]: You could change the traversal kind as well.

File: clang-tools-extra/clang-tidy/bugprone/BitCastPointersCheck.cpp [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+namespace clang::tidy::bugprone {
+
+void BitCastPointersCheck::registerMatchers(MatchFinder *Finder) {
+  auto IsPointerType = refersToType(qualType(isAnyPointer()));
+  Finder->addMatcher(callExpr(callee(functionDecl(allOf(
+                                  hasName("::std::bit_cast"),
+                                  hasTemplateArgument(0, IsPointerType),
+                                  hasTemplateArgument(1, IsPointerType)))))
+                         .bind("x"),
+                     this);
```
[the remark this answers is not in the document]
Comment (carlosgalvezp) [reply clang-tools-extra/clang-tidy/bugprone/BitCastPointersCheck.cpp:23]: Yes, that's correct. I think maybe this can be added as an extension in a follow-up patch if needed. Right now we cover the most common pitfall; if people want to workaround the check it will always be possible to do that one way or another.

File: clang-tools-extra/clang-tidy/bugprone/BitCastPointersCheck.cpp [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+}
+
+void BitCastPointersCheck::check(const MatchFinder::MatchResult &Result) {
+  if (const auto *MatchedDecl = Result.Nodes.getNodeAs<CallExpr>("x"))
+    diag(MatchedDecl->getBeginLoc(),
+         "do not use std::bit_cast on pointers; use it on values instead");
```
[the remark this answers is not in the document]
Comment (carlosgalvezp) [reply clang-tools-extra/clang-tidy/bugprone/BitCastPointersCheck.cpp:29]: I'm not sure I follow, how should I change the code?

File: clang-tools-extra/clang-tidy/bugprone/BitCastPointersCheck.cpp [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+  Finder->addMatcher(callExpr(callee(functionDecl(allOf(
+                                  hasName("::std::bit_cast"),
+                                  hasTemplateArgument(0, IsPointerType),
+                                  hasTemplateArgument(1, IsPointerType)))))
+                         .bind("x"),
+                     this);
+}
+
+void BitCastPointersCheck::check(const MatchFinder::MatchResult &Result) {
+  if (const auto *MatchedDecl = Result.Nodes.getNodeAs<CallExpr>("x"))
```
[the remark this answers is not in the document]
Comment (carlosgalvezp) [reply clang-tools-extra/clang-tidy/bugprone/BitCastPointersCheck.cpp:27]: In general it's good practice to check for null pointers, it's done in many tidy checks and is cheap to do. Node that the `getNodeAs` can return a nullptr even if there is a match, but it was not possible to cast to the chosen type.

File: clang-tools-extra/clang-tidy/bugprone/BitCastPointersCheck.h (hunk 2 above)
Comment (carlosgalvezp) [reply clang-tools-extra/clang-tidy/bugprone/BitCastPointersCheck.h:30]: I'm not familiar with traversal kinds, how should I change the code? What benefits would it bring? (Should the default traversal kind be changed in the add_new_check.py script?)

File: clang-tools-extra/clang-tidy/bugprone/BitCastPointersCheck.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (5chmidti) [reply clang-tools-extra/clang-tidy/bugprone/BitCastPointersCheck.cpp:23]: Sounds good

File: clang-tools-extra/clang-tidy/bugprone/BitCastPointersCheck.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (5chmidti) [reply clang-tools-extra/clang-tidy/bugprone/BitCastPointersCheck.cpp:27]: > Note that the `getNodeAs` can return a nullptr even if there is a match, but it was not possible to cast to the chosen type.

true

> it's done in many tidy checks and is cheap to do

I think some of the newer ones don't do that, but yes, it is cheap.

File: clang-tools-extra/clang-tidy/bugprone/BitCastPointersCheck.h (hunk 2 above)
Comment (5chmidti) [reply clang-tools-extra/clang-tidy/bugprone/BitCastPointersCheck.h:30]: Actually, if it were changed, template instantiations are ignored, so this shouldn't be done in this case, my bad.

TLDR: The `IgnoreUnlessSpelledInSource` traversal will ignore everything that the compiler will have to generate. E.g., implicit casts, generated special member functions, template instantiations, lambda structs,.... This allows for simpler matchers because they may not have to consider implicit casts, but can also improve performance because less analysis will be done.

> Should the default traversal kind be changed in the add_new_check.py script?

IMO no, because the default case is checking everything and including every AST node.

File: clang-tools-extra/clang-tidy/bugprone/BitCastPointersCheck.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (5chmidti) [reply clang-tools-extra/clang-tidy/bugprone/BitCastPointersCheck.cpp:29]: Steaming (`<<`) a source range into the diagnostic will underline the whole range:

```
18:16: warning: do not use std::bit_cast on pointers; use it on values instead [bugprone-bit-cast-pointers]
   18 |   float bad = *std::bit_cast<float*>(&x); // UB, but looks safe due to std::bit_cast
      |                ^
```

vs

```
18:16: warning: do not use std::bit_cast on pointers; use it on values instead [bugprone-bit-cast-pointers]
   18 |   float bad = *std::bit_cast<float*>(&x); // UB, but looks safe due to std::bit_cast
      |                ^~~~~~~~~~~~~~~~~~~~~~~~~
```

```c++
diag(MatchedDecl->getBeginLoc(),
         "do not use std::bit_cast on pointers; use it on values instead")
        << MatchedDecl.getSourceRange();
```

In case of FixItHint's, their replacement/deletion ranges are also highlighted in this way.

File: clang-tools-extra/test/clang-tidy/checkers/bugprone/bit-cast-pointers.cpp [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+}
+}
+
+void pointer2pointer()
+{
+  int x{};
+  float bad = *std::bit_cast<float*>(&x); // UB, but looks safe due to std::bit_cast
+  // CHECK-MESSAGES: :[[@LINE-1]]:16: warning: do not use std::bit_cast on pointers; use it on values instead [bugprone-bit-cast-pointers]
+  float good = std::bit_cast<float>(x);   // Well-defined
+}
```
Comment (5chmidti) [inline clang-tools-extra/test/clang-tidy/checkers/bugprone/bit-cast-pointers.cpp:21]: Could you add
```c++
  using IntPtr = int*;
  using FloatPtr = float*;
  IntPtr i;
  float bad2 = *std::bit_cast<FloatPtr>(i);
```
to the test as well? It passes, but aliased types should be tested as well.

**Comment by carlosgalvezp** [comment]
Thanks for the clarifications, will address asap! 

It now comes to mind that we probably also want to check `memcpy(ptr, ptr)`, which is equivalent to `bit_cast`. In that case I wonder if the check name still holds or it should be named something else?

**Comment by carlosgalvezp** [comment]
Friendly ping, let me know if there's anything else that should be addressed!

**Review by 5chmidti — COMMENTED** [review]
> It now comes to mind that we probably also want to check `memcpy(ptr, ptr)`, which is equivalent to `bit_cast`. In that case I wonder if the check name still holds or it should be named something else?

Yeah, with the addition of `memcpy`, the `bit` part of the name makes a less sense.
Maybe
- `bugprone-pointer-cast[ing]`
- `bugprone-cast[ing]-between-pointers`

File: clang-tools-extra/clang-tidy/bugprone/BitCastPointersCheck.cpp [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+                              hasDeclaration(functionDecl(hasName("::memcpy"))))
+                         .bind("memcpy"),
+                     this);
+}
+
+void BitCastPointersCheck::check(const MatchFinder::MatchResult &Result) {
+  if (const auto *MatchedDecl = Result.Nodes.getNodeAs<CallExpr>("bit_cast"))
+    diag(MatchedDecl->getBeginLoc(),
+         "do not use std::bit_cast to cast between pointers")
+        << MatchedDecl->getSourceRange();
```
[the remark this answers is not in the document]
Comment (carlosgalvezp) [reply clang-tools-extra/clang-tidy/bugprone/BitCastPointersCheck.cpp:38]: Then I have to add braces, though. Replaced with `else if` which should achieve the same goal.

**Comment by carlosgalvezp** [comment]
> > It now comes to mind that we probably also want to check `memcpy(ptr, ptr)`, which is equivalent to `bit_cast`. In that case I wonder if the check name still holds or it should be named something else?
> 
> Yeah, with the addition of `memcpy`, the `bit` part of the name makes a less sense. Maybe
> 
> * `bugprone-pointer-cast[ing]`
> * `bugprone-cast[ing]-between-pointers`

Hmm, I think "pointer cast" is too ambiguous, it could warn about _any_ type of pointer cast via e.g. `reinterpret_cast`, `const_cast`, etc. I want to put emphasis on the act of _copying the bytes_ of the pointer into another pointer to perform such a cast. In that sense I think "bit_cast" is still a more accurate name.

**Comment by carlosgalvezp** [comment]
What about `bugprone-bitwise-pointer-copy`?

**Comment by 5chmidti** [comment]
> What about `bugprone-bitwise-pointer-copy` @5chmidti ? `bugprone-bitwise-pointer-cast` could work as well

Both sound good to me.

**Comment by carlosgalvezp** [comment]
> Both sound good to me.

Great, updated! Let me know there's anything else to fix or if we can merge.

File: clang-tools-extra/clang-tidy/bugprone/BitwisePointerCastCheck.cpp [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+namespace clang::tidy::bugprone {
+
+void BitwisePointerCastCheck::registerMatchers(MatchFinder *Finder) {
+  auto IsPointerType = refersToType(qualType(isAnyPointer()));
+  Finder->addMatcher(callExpr(hasDeclaration(functionDecl(allOf(
+                                  hasName("::std::bit_cast"),
+                                  hasTemplateArgument(0, IsPointerType),
+                                  hasTemplateArgument(1, IsPointerType)))))
+                         .bind("bit_cast"),
+                     this);
```
Comment (5chmidti) [inline clang-tools-extra/clang-tidy/bugprone/BitwisePointerCastCheck.cpp:23]: `std::bit_cast` only exists in C++20>=, so this matcher can be moved behind a condition

**Review by 5chmidti — APPROVED** [review]
> Let me know there's anything else to fix or if we can merge.

Maybe give it like a day if someone wants to chime in.
